aten/src/ATen/native/cuda/UpSample.cuh

```diff
@@ -159,21 +169,32 @@ __device__ __forceinline__ static int nearest_neighbor_exact_compute_source_inde
 // see NOTE [ Nearest neighbor upsampling kernel implementation ]
 __device__ __forceinline__ static int nearest_neighbor_bw_compute_source_index(
     const float scale,
-    int dst_index,
-    int output_size) {
-  // Equivalent to buggy OpenCV INTER_NEAREST
-  // We keep this method for BC and consider as deprecated.
-  // See nearest_neighbor_exact_bw_compute_source_index as replacement
-  const int src_index =
-      min(static_cast<int>(ceilf(dst_index * scale)), output_size);
-  return src_index;
+    int dst_index,  // grad_input index
+    int output_size,  // grad_output size
+    int input_size  // grad_input size
+) {
+  if (output_size == input_size) {
+    // no matter scale_factor, simply copy as for CPU
+    return dst_index;
+  } else if (output_size == 2 * input_size) {
+    // no matter scale_factor, shift input index as for CPU
+    return dst_index << 1;
+  } else {
```

Comment:
So IIUC the crux of this PR are these newly added lines, and the goal here is to align the CUDA backward pass with the CPU backward pass. I'll refer to those lines as the "fast path".

One thing that we'd need to make sure before merging is: is the fast path actually **more correct** than the slow path? And why is the fast path different from the slow path in terms of results - we would expect them to be the same.

It's important to answer that because there are 2 ways to align CUDA and CPU here: either make CUDA the same as CPU (this PR), or make CPU the same as CUDA. If the fast path is not strictly correct, then a more appropriate solution would be to remove it from the CPU part instead of adding it to CUDA.

Reply:
Thanks for raising this duality in a way to fix the problem.

### Context

Actually, there are two bugs in the interpolate op:
1) output size is incorrectly computed if using scale (https://github.com/pytorch/pytorch/issues/62396)
2) "nearest" mode is incorrect, same as opencv. That's why "nearest-exact" mode was introduced similary to opencv (https://github.com/pytorch/pytorch/pull/64501).
3) Two fast paths for `output_size == input_size` and `output_size == 2 * input_size` are present in "nearest" mode, cpu forward pass, index computation methods.
  - first fast-path: https://github.com/pytorch/pytorch/blob/15686950b71e27261ad25cd8ce45763b7f2d6eb0/aten/src/ATen/native/UpSample.h#L340-L342
  - second fast-path: https://github.com/pytorch/pytorch/blob/15686950b71e27261ad25cd8ce45763b7f2d6eb0/aten/src/ATen/native/UpSample.h#L343-L345
 
### Which code paths are correct ?

#### The fast path for `output_size == input_size`

Comparision of input indices computed for nearest mode (not nearest-exact where there are no fast paths) for both fast and slow paths gives the following. For example, 
```python
input_size = 4
s = 1.123  # scale here is taken such that output size is correct
output_size = int(input_size * s)

input_indices_1 = []
input_indices_2 = []

if output_size == input_size:
    for output_index in range(output_size):        
        input_index = min(int(output_index / s), input_size-1)  # buggy nearest mode
        input_indices_1.append(input_index)
        input_indices_2.append(output_index)

print(input_indices_1, input_indices_2)
> ([0, 0, 1, 2], [0, 1, 2, 3])
```
Here `input_indices_1` are computed with slow-path and `input_indices_2` are computed with the 1st fast-path.
Evidently, `input_indices_1` looks incorrect. By the way, "nearest-exact" mode also gives correct indices like `input_indices_2`.

The 1st fast path (for `output_size == input_size`) is also present in opencv: https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/resize.cpp#L4092-L4097

So, we can assume that the fast path is correct for buggy "nearest" mode.

#### The fast path for `output_size == 2 * input_size`

Similar indices computation:
```python
input_size = 4
s = 2.1234  # scale here is taken such that output size is correct
output_size = int(input_size * s)

input_indices_1 = []
input_indices_2 = []
input_indices_3 = []

for output_index in range(output_size):        
    input_index = min(int(output_index / s), input_size-1)  # buggy nearest mode
    input_index_2 = min(int(output_index / s + 0.5), input_size-1)  # nearest-exact mode
    input_indices_1.append(input_index)
    input_indices_2.append(output_index >> 1)
    input_indices_3.append(input_index_2)

print(input_indices_1, input_indices_2, input_indices_3)
> [0, 0, 0, 1, 1, 2, 2, 3] [0, 0, 1, 1, 2, 2, 3, 3] [0, 0, 1, 1, 2, 2, 3, 3]
```
 can show that fast-path indices (`input_indices_2`) looks correct (same as nearest-exact mode, `input_indices_3`) vs slow-path indices `input_indices_1`.

However, the output wont match opencv result (which is also buggy).
```python
import cv2
import numpy as np

w = 4
inpt = np.arange(3 * w * w, dtype="float32").reshape(3, w, w).transpose((1, 2, 0))
s = 2.1234
out = cv2.resize(inpt, (0, 0), fx=s, fy=s, interpolation=cv2.INTER_NEAREST)
out = out.transpose((2, 0, 1))
print(out.shape)
out[0, 0, :]
> array([0., 0., 0., 1., 1., 2., 2., 3.], dtype=float32)
```

Another observation is that pytorch output for other similar larger scales like `s=N + 0.1234` will match opencv, "nearest" mode.
Thus, I would say that this fast-path (for `output_size == 2 * input_size`) is incorrect here if we keep opencv as a reference for "nearest" mode.

### What to do next ? 

Nicolas, I agree with you that 

> there are 2 ways to align CUDA and CPU here: either make CUDA the same as CPU (this PR), or make CPU the same as CUDA. If the fast path is not strictly correct, then a more appropriate solution would be to remove it from the CPU part instead of adding it to CUDA.

I think we should keep everywhere (cpu/cuda) the first fast-path (`output_size == input_size`) and thus fix cuda grads for this case.

I do not know if we can simply remove the second fast-path (`output_size == 2 * input_size`) from cpu forward/backward passes without BC-breaking considerations. 

@jbschlosser @ngimel can you please give your feedback here?

Reply:
Thanks a lot for the detailed analysis @vfdev-5 . Just summarizing a discussion we just had offline with @vfdev-5 below.

(**TL;DR**: use first fast path everywhere, don't use second fast path anywhere)

- First fast path == correct == what OpenCV does    => **easy decision**, we use the correct version which also aligns with openCV (ie. use fast path)

- Second fast path (`output_size == 2 * input_size`) == correct != what OpenCV does  => we need to decide whether we want the "correct" way or align with openCV. Pytorch is aligned with openCV for any other scale that isn't 2, e.g. 3 or 4 would be aligned with openCV. So it make sense to keep alignment between pytorch and opencv for the `output_size == 2 * input_size` case as well, regardless of correctness (they're both incorrect, that's why nearest-exact exists). So we probably should **not** use the fast-path here.


Both of these can pass as bugfixes IMHO and so we don't have to worry about BC-breaking considerations.